File: code/dart/grpc/server/lib/services/task_service.dart
import 'dart:async';
import 'package:grpc/grpc.dart';
import 'package:uuid/uuid.dart';

// Manual proto message definitions (normally generated)
class Task {
  String id;
  String title;
  String description;
  String status;
  String priority;
  List<String> tags;
  String createdBy;
  String assignedTo;
  Timestamp? createdAt;
  Timestamp? updatedAt;
  Timestamp? dueDate;
  Timestamp? completedAt;

  Task({
    required this.id,
    required this.title,
    this.description = '',
    this.status = 'TASK_STATUS_PENDING',
    this.priority = 'TASK_PRIORITY_MEDIUM',
    List<String>? tags,
    this.createdBy = 'system',
    this.assignedTo = '',
    this.createdAt,
    this.updatedAt,
    this.dueDate,
    this.completedAt,
  }) : tags = tags ?? [];

  Map<String, dynamic> toProto() => {
        'id': id,
        'title': title,
        'description': description,
        'status': status,
        'priority': priority,
        'tags': tags,
        'created_by': createdBy,
        'assigned_to': assignedTo,
        'created_at': createdAt?.toProto(),
        'updated_at': updatedAt?.toProto(),
        'due_date': dueDate?.toProto(),
        'completed_at': completedAt?.toProto(),
      };
}

class Timestamp {
  final int seconds;
  final int nanos;

  Timestamp({required this.seconds, this.nanos = 0});

  factory Timestamp.fromDateTime(DateTime dt) {
    final ms = dt.millisecondsSinceEpoch;
    return Timestamp(
      seconds: ms ~/ 1000,
      nanos: (ms % 1000) * 1000000,
    );
  }

  Map<String, dynamic> toProto() => {
        'seconds': seconds,
        'nanos': nanos,
      };
}

class ListTasksRequest {
  int? pageSize;
  String? pageToken;
  String? status;
  String? assignedTo;
  List<String> tags;
  String? sortOrder;

  ListTasksRequest({
    this.pageSize,
    this.pageToken,
    this.status,
    this.assignedTo,
    List<String>? tags,
    this.sortOrder,
  }) : tags = tags ?? [];
}

class GetTaskRequest {
  final String id;
  GetTaskRequest(this.id);
}

class CreateTaskRequest {
  final Task task;
  CreateTaskRequest(this.task);
}

class UpdateTaskRequest {
  final Task task;
  final List<String> updateMask;
  UpdateTaskRequest(this.task, this.updateMask);
}

class DeleteTaskRequest {
  final String id;
  DeleteTaskRequest(this.id);
}

class WatchTasksRequest {
  List<String> taskIds;
  bool watchAll;
  String? assignedTo;

  WatchTasksRequest({
    List<String>? taskIds,
    this.watchAll = false,
    this.assignedTo,
  }) : taskIds = taskIds ?? [];
}

class TaskEvent {
  String eventType;
  Task task;
  Timestamp timestamp;

  TaskEvent({
    required this.eventType,
    required this.task,
    required this.timestamp,
  });

  Map<String, dynamic> toProto() => {
        'event_type': eventType,
        'task': task.toProto(),
        'timestamp': timestamp.toProto(),
      };
}

class TaskServiceImpl extends Service {
  final Map<String, Task> _tasks = {};
  final List<StreamController<TaskEvent>> _watchers = [];

  TaskServiceImpl() {
    _initializeSampleData();
  }

  void _initializeSampleData() {
    final sampleTasks = [
      Task(
        id: const Uuid().v4(),
        title: 'Implement Dart gRPC server',
        description: 'Create gRPC server in Dart',
        status: 'TASK_STATUS_IN_PROGRESS',
        priority: 'TASK_PRIORITY_HIGH',
        tags: ['grpc', 'dart', 'server'],
        createdBy: 'system',
        assignedTo: 'developer',
        createdAt: Timestamp.fromDateTime(DateTime.now()),
        updatedAt: Timestamp.fromDateTime(DateTime.now()),
      ),
      Task(
        id: const Uuid().v4(),
        title: 'Test streaming functionality',
        description: 'Implement and test bidirectional streaming',
        status: 'TASK_STATUS_PENDING',
        priority: 'TASK_PRIORITY_MEDIUM',
        tags: ['grpc', 'streaming', 'testing'],
        createdBy: 'system',
        assignedTo: 'developer',
        createdAt: Timestamp.fromDateTime(DateTime.now()),
        updatedAt: Timestamp.fromDateTime(DateTime.now()),
      ),
    ];

    for (final task in sampleTasks) {
      _tasks[task.id] = task;
    }
  }

  // Server streaming: List tasks
  Stream<Task> listTasks(ServiceCall call, ListTasksRequest request) async* {
    var tasks = _tasks.values.toList();

    // Apply filters
    if (request.status != null && request.status != 'TASK_STATUS_UNSPECIFIED') {
      tasks = tasks.where((task) => task.status == request.status).toList();
    }

    if (request.assignedTo != null) {
      tasks = tasks
          .where((task) => task.assignedTo == request.assignedTo)
          .toList();
    }

    if (request.tags.isNotEmpty) {
      tasks = tasks.where((task) {
        return request.tags.every((tag) => task.tags.contains(tag));
      }).toList();
    }

    // Stream tasks to client
    for (final task in tasks) {
      yield task;
    }
  }

  // Unary call: Get single task
  Future<Task> getTask(ServiceCall call, GetTaskRequest request) async {
    final task = _tasks[request.id];
    if (task == null) {
      throw GrpcError.notFound('Task with ID ${request.id} not found');
    }
    return task;
  }

  // Unary call: Create task
  Future<Task> createTask(ServiceCall call, CreateTaskRequest request) async {
    final task = request.task;

    if (task.title.isEmpty) {
      throw GrpcError.invalidArgument('Task title is required');
    }

    final newTask = Task(
      id: const Uuid().v4(),
      title: task.title,
      description: task.description,
      status: task.status.isNotEmpty ? task.status : 'TASK_STATUS_PENDING',
      priority:
          task.priority.isNotEmpty ? task.priority : 'TASK_PRIORITY_MEDIUM',
      tags: task.tags,
      createdBy: task.createdBy.isNotEmpty ? task.createdBy : 'system',
      assignedTo: task.assignedTo,
      createdAt: Timestamp.fromDateTime(DateTime.now()),
      updatedAt: Timestamp.fromDateTime(DateTime.now()),
      dueDate: task.dueDate,
    );

    _tasks[newTask.id] = newTask;

    // Notify watchers
    _notifyWatchers('EVENT_TYPE_CREATED', newTask);

    return newTask;
  }

  // Unary call: Update task
  Future<Task> updateTask(ServiceCall call, UpdateTaskRequest request) async {
    final task = request.task;

    if (task.id.isEmpty) {
      throw GrpcError.invalidArgument('Task ID is required');
    }

    final existingTask = _tasks[task.id];
    if (existingTask == null) {
      throw GrpcError.notFound('Task with ID ${task.id} not found');
    }

    // Apply updates
    final updatedTask = Task(
      id: existingTask.id,
      title: request.updateMask.contains('title') && task.title.isNotEmpty
          ? task.title
          : existingTask.title,
      description: request.updateMask.contains('description')
          ? task.description
          : existingTask.description,
      status: request.updateMask.contains('status') && task.status.isNotEmpty
          ? task.status
          : existingTask.status,
      priority:
          request.updateMask.contains('priority') && task.priority.isNotEmpty
              ? task.priority
              : existingTask.priority,
      tags: request.updateMask.contains('tags') ? task.tags : existingTask.tags,
      createdBy: existingTask.createdBy,
      assignedTo: request.updateMask.contains('assigned_to')
          ? task.assignedTo
          : existingTask.assignedTo,
      createdAt: existingTask.createdAt,
      updatedAt: Timestamp.fromDateTime(DateTime.now()),
      dueDate: request.updateMask.contains('due_date')
          ? task.dueDate
          : existingTask.dueDate,
      completedAt: task.status == 'TASK_STATUS_COMPLETED' &&
              existingTask.status != 'TASK_STATUS_COMPLETED'
          ? Timestamp.fromDateTime(DateTime.now())
          : existingTask.completedAt,
    );

    _tasks[task.id] = updatedTask;

    // Notify watchers
    _notifyWatchers('EVENT_TYPE_UPDATED', updatedTask);

    return updatedTask;
  }

  // Unary call: Delete task
  Future<Empty> deleteTask(ServiceCall call, DeleteTaskRequest request) async {
    final task = _tasks[request.id];
    if (task == null) {
      throw GrpcError.notFound('Task with ID ${request.id} not found');
    }

    _tasks.remove(request.id);

    // Notify watchers
    _notifyWatchers('EVENT_TYPE_DELETED', task);

    return Empty();
  }

  // Bidirectional streaming: Watch tasks
  Stream<TaskEvent> watchTasks(
      ServiceCall call, Stream<WatchTasksRequest> requests) async* {
    final controller = StreamController<TaskEvent>();
    _watchers.add(controller);

    // Handle incoming requests
    final subscription = requests.listen((request) {
      if (request.watchAll) {
        // Send current tasks
        for (final task in _tasks.values) {
          controller.add(TaskEvent(
            eventType: 'EVENT_TYPE_CREATED',
            task: task,
            timestamp: Timestamp.fromDateTime(DateTime.now()),
          ));
        }
      } else if (request.taskIds.isNotEmpty) {
        for (final id in request.taskIds) {
          final task = _tasks[id];
          if (task != null) {
            controller.add(TaskEvent(
              eventType: 'EVENT_TYPE_CREATED',
              task: task,
              timestamp: Timestamp.fromDateTime(DateTime.now()),
            ));
          }
        }
      }
    });

    // Cleanup on cancel
    call.onCancel.then((_) {
      subscription.cancel();
      controller.close();
      _watchers.remove(controller);
    });

    // Stream events
    yield* controller.stream;
  }

  void _notifyWatchers(String eventType, Task task) {
    final event = TaskEvent(
      eventType: eventType,
      task: task,
      timestamp: Timestamp.fromDateTime(DateTime.now()),
    );

    for (final watcher in _watchers) {
      if (!watcher.isClosed) {
        watcher.add(event);
      }
    }

    // Clean up closed watchers
    _watchers.removeWhere((w) => w.isClosed);
  }
}

// Placeholder for protocol buffer Empty message
class Empty {}

File: code/javascript/grpc/server/services/taskService.js
import grpc from '@grpc/grpc-js';
import { v4 as uuidv4 } from 'uuid';

export class TaskServiceImpl {
  constructor() {
    this.tasks = new Map();
    this.watchers = new Set();
    this.initializeSampleData();
  }

  initializeSampleData() {
    const sampleTasks = [
      {
        id: uuidv4(),
        title: 'Implement gRPC server',
        description: 'Create gRPC server in JavaScript',
        status: 'TASK_STATUS_IN_PROGRESS',
        priority: 'TASK_PRIORITY_HIGH',
        tags: ['grpc', 'javascript', 'server'],
        created_by: 'system',
        assigned_to: 'developer',
        created_at: this.getCurrentTimestamp(),
        updated_at: this.getCurrentTimestamp()
      },
      {
        id: uuidv4(),
        title: 'Test bidirectional streaming',
        description: 'Implement and test Watch functionality',
        status: 'TASK_STATUS_PENDING',
        priority: 'TASK_PRIORITY_MEDIUM',
        tags: ['grpc', 'streaming', 'testing'],
        created_by: 'system',
        assigned_to: 'developer',
        created_at: this.getCurrentTimestamp(),
        updated_at: this.getCurrentTimestamp()
      }
    ];
    
    sampleTasks.forEach(task => {
      this.tasks.set(task.id, task);
    });
  }

  getCurrentTimestamp() {
    const now = new Date();
    return {
      seconds: Math.floor(now.getTime() / 1000),
      nanos: (now.getTime() % 1000) * 1000000
    };
  }

  // Server streaming: List tasks
  listTasks(call) {
    const request = call.request;
    let tasks = Array.from(this.tasks.values());
    
    // Apply filters
    if (request.status && request.status !== 'TASK_STATUS_UNSPECIFIED') {
      tasks = tasks.filter(task => task.status === request.status);
    }
    
    if (request.assigned_to) {
      tasks = tasks.filter(task => task.assigned_to === request.assigned_to);
    }
    
    if (request.tags && request.tags.length > 0) {
      tasks = tasks.filter(task =>
        request.tags.every(tag => task.tags.includes(tag))
      );
    }
    
    // Apply sorting
    tasks = this.sortTasks(tasks, request.sort_order);
    
    // Apply pagination
    const pageSize = request.page_size || 20;
    const startIndex = request.page_token ? parseInt(request.page_token) : 0;
    const paginatedTasks = tasks.slice(startIndex, startIndex + pageSize);
    
    // Stream tasks to client
    paginatedTasks.forEach(task => {
      call.write(task);
    });
    
    call.end();
  }

  // Unary call: Get single task
  getTask(call, callback) {
    const { id } = call.request;
    const task = this.tasks.get(id);
    
    if (!task) {
      return callback({
        code: grpc.status.NOT_FOUND,
        message: `Task with ID ${id} not found`
      });
    }
    
    callback(null, task);
  }

  // Unary call: Create task
  createTask(call, callback) {
    const { task } = call.request;
    
    if (!task.title) {
      return callback({
        code: grpc.status.INVALID_ARGUMENT,
        message: 'Task title is required'
      });
    }
    
    const newTask = {
      id: uuidv4(),
      title: task.title,
      description: task.description || '',
      status: task.status || 'TASK_STATUS_PENDING',
      priority: task.priority || 'TASK_PRIORITY_MEDIUM',
      tags: task.tags || [],
      created_by: task.created_by || 'system',
      assigned_to: task.assigned_to || '',
      created_at: this.getCurrentTimestamp(),
      updated_at: this.getCurrentTimestamp(),
      due_date: task.due_date || null,
      completed_at: null
    };
    
    this.tasks.set(newTask.id, newTask);
    
    // Notify watchers
    this.notifyWatchers('EVENT_TYPE_CREATED', newTask);
    
    callback(null, newTask);
  }

  // Unary call: Update task
  updateTask(call, callback) {
    const { task, update_mask } = call.request;
    
    if (!task.id) {
      return callback({
        code: grpc.status.INVALID_ARGUMENT,
        message: 'Task ID is required'
      });
    }
    
    const existingTask = this.tasks.get(task.id);
    if (!existingTask) {
      return callback({
        code: grpc.status.NOT_FOUND,
        message: `Task with ID ${task.id} not found`
      });
    }
    
    // Apply updates based on update_mask or all provided fields
    const updatedTask = { ...existingTask };
    
    if (update_mask && update_mask.length > 0) {
      // Update only specified fields
      update_mask.forEach(field => {
        if (task[field] !== undefined) {
          updatedTask[field] = task[field];
        }
      });
    } else {
      // Update all provided fields
      Object.keys(task).forEach(key => {
        if (task[key] !== undefined && key !== 'id' && key !== 'created_at') {
          updatedTask[key] = task[key];
        }
      });
    }
    
    updatedTask.updated_at = this.getCurrentTimestamp();
    
    // Handle status change to completed
    if (updatedTask.status === 'TASK_STATUS_COMPLETED' && 
        existingTask.status !== 'TASK_STATUS_COMPLETED') {
      updatedTask.completed_at = this.getCurrentTimestamp();
    }
    
    this.tasks.set(task.id, updatedTask);
    
    // Notify watchers
    this.notifyWatchers('EVENT_TYPE_UPDATED', updatedTask);
    
    callback(null, updatedTask);
  }

  // Unary call: Delete task
  deleteTask(call, callback) {
    const { id } = call.request;
    const task = this.tasks.get(id);
    
    if (!task) {
      return callback({
        code: grpc.status.NOT_FOUND,
        message: `Task with ID ${id} not found`
      });
    }
    
    this.tasks.delete(id);
    
    // Notify watchers
    this.notifyWatchers('EVENT_TYPE_DELETED', task);
    
    callback(null, {});
  }

  // Bidirectional streaming: Watch tasks
  watchTasks(call) {
    // Add this call to watchers
    this.watchers.add(call);
    
    // Handle incoming requests
    call.on('data', (request) => {
      console.log('Watch request received:', request);
      
      // Send current tasks matching the filter
      if (request.watch_all) {
        this.tasks.forEach(task => {
          call.write({
            event_type: 'EVENT_TYPE_CREATED',
            task: task,
            timestamp: this.getCurrentTimestamp()
          });
        });
      } else if (request.task_ids && request.task_ids.length > 0) {
        request.task_ids.forEach(id => {
          const task = this.tasks.get(id);
          if (task) {
            call.write({
              event_type: 'EVENT_TYPE_CREATED',
              task: task,
              timestamp: this.getCurrentTimestamp()
            });
          }
        });
      }
    });
    
    // Handle client disconnect
    call.on('end', () => {
      this.watchers.delete(call);
      call.end();
    });
    
    // Handle errors
    call.on('error', (err) => {
      console.error('Watch stream error:', err);
      this.watchers.delete(call);
    });
  }

  // Helper: Notify all watchers of task changes
  notifyWatchers(eventType, task) {
    const event = {
      event_type: eventType,
      task: task,
      timestamp: this.getCurrentTimestamp()
    };
    
    this.watchers.forEach(watcher => {
      try {
        watcher.write(event);
      } catch (err) {
        console.error('Failed to notify watcher:', err);
        this.watchers.delete(watcher);
      }
    });
  }

  // Helper: Sort tasks
  sortTasks(tasks, sortOrder) {
    switch (sortOrder) {
      case 'SORT_ORDER_CREATED_AT_ASC':
        return tasks.sort((a, b) => a.created_at.seconds - b.created_at.seconds);
      case 'SORT_ORDER_CREATED_AT_DESC':
        return tasks.sort((a, b) => b.created_at.seconds - a.created_at.seconds);
      case 'SORT_ORDER_DUE_DATE_ASC':
        return tasks.sort((a, b) => {
          if (!a.due_date) return 1;
          if (!b.due_date) return -1;
          return a.due_date.seconds - b.due_date.seconds;
        });
      case 'SORT_ORDER_DUE_DATE_DESC':
        return tasks.sort((a, b) => {
          if (!a.due_date) return 1;
          if (!b.due_date) return -1;
          return b.due_date.seconds - a.due_date.seconds;
        });
      case 'SORT_ORDER_PRIORITY_ASC':
        return tasks.sort((a, b) => this.priorityValue(a.priority) - this.priorityValue(b.priority));
      case 'SORT_ORDER_PRIORITY_DESC':
        return tasks.sort((a, b) => this.priorityValue(b.priority) - this.priorityValue(a.priority));
      default:
        return tasks;
    }
  }

  priorityValue(priority) {
    const values = {
      'TASK_PRIORITY_LOW': 1,
      'TASK_PRIORITY_MEDIUM': 2,
      'TASK_PRIORITY_HIGH': 3,
      'TASK_PRIORITY_CRITICAL': 4
    };
    return values[priority] || 2;
  }
}